### ISSUE TITLE
Burned-in subtitles show at the incorrect time

### ISSUE THREAD

**Issue by bachya** [donmelton/video_transcoding#131]
I have an MKV of The Phantom Menace with a "forced" subtitle track that `transcode-video` doesn't automatically detect. Therefore, I attempted to use the `--burn-subtitle` parameter to force the track in:

```
transcode-video --crop 130:132:0:0 --burn-subtitle 38 /path/to/tpm.mkv
```

...however, in the final transcoded video, the subtitles appear too soon (in each case, they appear/disappear about 0.5 to 1 second before they should).

For whatever it's worth, VLC doesn't play this "forced" track automatically, either, but when it is manually selected in the source MKV (i.e., the MKV used _by_ `transcode-video`), the subtitles appear at the correct time.

I've attached the `transcode-video` log here in case it is of use. Appreciate your assistance!

[Star Wars Episode I - The Phantom Menace (1999).mkv.log.zip](https://github.com/donmelton/video_transcoding/files/802346/Star.Wars.Episode.I.-.The.Phantom.Menace.1999.mkv.log.zip)

**Comment by bachya** [comment]
Quick update: it appears that this particular Blu-ray (and, indeed, all of the Blu-rays in this set) have the desired "forced" track as a separate track; it isn't flagged by default. So, as a new test, I've remuxed the original MKV (using MKVToolNix) and set the "force" flag for that track; I'll run it through `transcode-video` again and see if anything changes (I'm hoping that this track is just "picked up" now...).

**Comment by donmelton** [comment]
@bachya Yes, the forced subtitle track on episodes in the "Star Wars" boxed set, at least those four which do have forced subtitles, is always the last subtitle track, except for "Episode IV - A New Hope" which is the second subtitle track. BTW, "Episode III - Revenge of the Sith" and "Episode V - The Empire Strikes Back" do not have forced subtitles.

As to your bug, that's very likely a problem with the way your original rips were muxed. The current version of HandBrake has a problem with offset delays in certain MKV files.

Let me know how it goes with your test of the remuxed files.

**Comment by bachya** [comment]
Thanks for the comments, @donmelton!

Unfortunately, remuxing the source MKV didn't help the end result – the same effect, wherein the newly transcoded MKV's now-burned-in subtitles appear 0.5-1 seconds late, is still occurring.

For reference, here's what appears at 32:06 for both MKVs:

**Original (i.e., built from MakeMKV)**

<img width="1280" alt="original" src="https://cloud.githubusercontent.com/assets/47216/23364317/db2e1946-fcbb-11e6-87a5-85a148bfed7b.png">

**Transcoded (i.e., using `transcode-video`)**

<img width="1280" alt="transcoded" src="https://cloud.githubusercontent.com/assets/47216/23364343/f4f3653e-fcbb-11e6-84ce-8a6801677a02.png">

(I'm sure you're already familiar with this film, but in case not, the original MKV is showing that forced subtitle track at the correct time; the transcoded MKV's burned-in version of that subtitle track is late.)

**Comment by donmelton** [comment]
@bachya Thanks for doing all this testing! And for the useful and demonstrative screen shots!

Yes, I'm very familiar with this film. :) I've transcoded all seven "Star Wars" Blu-ray Disc titles at least a dozen times. I just checked my most recent transcoding of "Episode I - The Phantom Menace" and I don't see this delay even after verifying with both VLC and `mpv`.

My original hypothesis still stands. I suspect there's something peculiar about your ripped MKV file, either the original or the remux, which is confusing HandBrake. Very possibly it's a delay offset on the subtitle tracks.

How did you rip the original disc? With MakeMKV?

**Comment by bachya** [comment]
>How did you rip the original disc? With MakeMKV?

Yes sir, I'm using MakeMKV v1.10.4 (running on Ubuntu 16.04.2 LTS). I didn't do anything "special" – I just loaded up the Blu-ray, picked the title I wanted, and started decoding it.

I know you're not responsible for supporting MakeMKV, but if it's okay, I'll share with you my settings; perhaps you can see something I'm doing wrong?

**Video**

* Default destination: Custom (`/media/media`)
* Minimum title length (seconds): 120

**IO**

* Read retry count: 5
* Read buffer size (MB): Auto
* Ask for single drive mode: <unchecked>

**Language**

* Interface language: eng : English
* Preferred language: eng : English

**Protection**

* DVD structure protection removal method: Auto
* Always create BD+ dumps: <unchecked>
* Custom Java executable location: <empty>

**Options**

* Enable UPnP: <checked>
* Use specific IP address: Auto
* Web server port: 51000

**General**

* MakeMKV data directory: `/home/bachya/.MakeMKV`
* Log debug messages: <unchecked>
* Allow contacting web server for updates: <checked>
* Expert mode: <checked>
* Show AV synchronization messages: <unchecked>

**Advanced**

* Default profile: Default
* Default selection rule: `-sel:all,+sel:(favlang|nolang|single),-sel:(havemulti|havecore),-sel:mvcvideo,=100:all,-10:favlang`
* dtsdecoderdll.dll: location: <empty>

**Comment by donmelton** [comment]
@bachya Thanks for the details! Yeah, that looks like a straightforward configuration for MakeMKV. So nothing odd there.

Do you have the `mediainfo` tool installed? If so, can you attach the output of `mediainfo --Full` on your original MKV file? Maybe something will stand out there. Thanks.

**Comment by bachya** [comment]
Absolutely!

[tpm.log.zip](https://github.com/donmelton/video_transcoding/files/804449/tpm.log.zip)

**Comment by donmelton** [comment]
@bachya Thanks! The good news is that none of your tracks -- video, audio or subtitles -- have delays. That bad news is that's a dead end for diagnosing the problem.

OK, I'm stumped. The only difference in my transcoding and yours of the same movie is that I used `HandBrakeCLI` version 1.0.1 and you used version 1.0.2.

So, I'll make a new transcoding using version 1.0.2 and see what happens. Stay tuned...

**Comment by bachya** [comment]
Thanks so much, @donmelton! Really appreciate your help.

**Comment by donmelton** [comment]
@bachya That's what I'm here for. :)

BTW, the other difference, of course, is that I'm using macOS and you're using Linux. One would hope that the Linux version doesn't differ all that much but you never know.

**Comment by bachya** [comment]
Understood; if you deem that 1.0.2 doesn't help, I'll go back to the MakeMKV folks and see if they can help.

**Comment by donmelton** [comment]
@bachya OK, the transcoding using `HandBrakeCLI` version 1.0.2 just finished and it's essentially identically my version using 1.0.1, meaning that the subtitles are in sync.

So... I'm out of ideas here. But let's summarize the two areas where the problem could be:

* MakeMKV is ripping the title incorrectly for you. This seems less likely after you sent me the diagnostic information from `mediainfo`.
* `HandBrakeCLI` is transcoding the title incorrectly for you. This seems to be the most likely cause. And I can only guess that I'm not seeing it with the same version because I'm using macOS and you're using Linux.

Honestly, I don't know what to suggest at this point. I am truly sorry that I can't diagnose this fully for you. :(

**Comment by bachya** [comment]
Thanks so much, @donmelton – appreciate the lengths to which you've gone.

Unfortunately, my Macbook doesn't have a disc drive, so I can't test the OS difference very easily. Going in the other direction, however, I have an idea.

I could create a container (either a VirtualBox virtual machine or a Docker container) that mimics my environment _exactly_ (same OS version, same version of MakeMKV, same version of all the tools that your stuff needs); then, I could deliver that to you and you'd be able to test that same environment yourself.

Interested? No pressure at all, but if so, let me know and I can get to work.

**Comment by donmelton** [comment]
@bachya I've never used Docker but that might be easier setting up VirtualBox again. I will consider it and let you know.

In the meantime, I suggest opening an issue with the HandBrake team about this.

**Comment by bachya** [comment]
One other question, @donmelton: my system currently only has `handbrake-cli` installed (i.e., I don't have the `handbrake` [GUI] package installed). That shouldn't matter, right?

**Comment by donmelton** [comment]
@bachya Correct, that shouldn't matter.

**Comment by bachya** [comment]
Submitted a bug report to the Handbrake folks: https://github.com/HandBrake/HandBrake/issues/601

**Comment by donmelton** [comment]
@bachya Cool. I'll leave this issue open for awhile while they review the problem.

**Comment by bachya** [comment]
Hi @donmelton – quick update: the issue is fixed (thanks to the Handbrake folks)!

For anyone finding this in the future, it turns out that the version of `handbrake-cli` stored in the Ubuntu Xenial repositories (0.10.2 as of this writing) is (a) out-of-date and (b) broken. I followed [the Handbrake team's instructions for installing the most up-to-date version](https://launchpad.net/~stebbins/+archive/ubuntu/handbrake-releases) (1.0.3 as of this writing) and, after a fresh transcode, we're in business!

For these particular Blu-rays, I'm not sure if the intermediate step – in which I used MKVToolNix to remux the original MKV and force the subtitle track – is necessary, but it can't hurt.

Thanks again for your help! I wouldn't have thought to go to the Handbrake folks unless you had suggested it.
